lib/routes/recovery-keys.js

```diff
@@ -0,0 +1,230 @@
+        log.begin('createRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const {recoveryKeyId, recoveryData} = request.payload
+
+        customs.check(request, 'createRecoveryKey')
+          .then(createRecoveryKey)
+          .then(emitMetrics)
+          .then(() => reply({}), reply)
+
+        function createRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.createRecoveryKey(uid, recoveryKeyId, recoveryData)
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.recoveryKey.created',
+            uid: uid
+          })
+
+          return request.emitMetricsEvent('recoveryKey.created', {uid: uid})
+            .then(() => ({}))
+        }
+      }
+    },
+    {
+      method: 'GET',
+      path: '/recoveryKeys/{recoveryKeyId}',
+      config: {
+        auth: {
+          strategy: 'sessionToken'
+        }
+      },
+      handler(request, reply) {
+        log.begin('getRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const recoveryKeyId = request.params.recoveryKeyId
+        let recoveryData
+
+        customs.check(request, 'getRecoveryKey')
+          .then(getRecoveryKey)
+          .then(() => reply({recoveryData}), reply)
+
+        function getRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.getRecoveryKey(uid, recoveryKeyId)
+            .then((res) => recoveryData = res.recoveryData)
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/reset/recoveryKeys',
+      config: {
+        auth: {
+          strategy: 'accountResetToken'
+        },
+        validate: {
+          query: {
+            keys: isA.boolean().optional()
+          },
+          payload: {
+            authPW: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            wrapKb: isA.string().min(64).max(64).regex(HEX_STRING).required(),
+            recoveryKeyId: isA.string().max(256).required(),
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Password.changeRecoveryKeys', request)
+        const authPW = request.payload.authPW
+        const wrapKb = request.payload.wrapKb
+        const recoveryKeyId = request.payload.recoveryKeyId
+        const uid = request.auth.credentials.uid
+
+        let authSalt, password, verifyHash, sessionToken, keyFetchToken, account
+
+        checkRecoveryKey()
+          .then(getAccount)
+          .then(passwordReset)
+          .then(deleteRecoveryKey)
+          .then(notifyAccount)
+          .then(createSessionToken)
+          .then(createKeyFetchToken)
+          .then(createResponse)
```

Comment:
As part of this process, should we automatically generate and return a *new* recovery key for the user so that the new recovery key can be displayed to the user w/o the need to manually set up a new recovery key? In the UI, it looks like the [user is sent directly to /reset_password_verified](https://github.com/mozilla/fxa-content-server/pull/6203/files#diff-ffc6d69aee72d476ab73f60dfed968e9R24) which gets the user into the account, but doesn't prepare them for the next time they need a recovery code.

Reply:
Unfortunately, auto regenerating a recovery key hasn't been completely flushed out yet. I am not sure if we can automatically generate one server-side since that was mentioned in the security review. 

Either way, there should be some sort of messaging to let the user know at the end that this key can not be used again and give them some options to generate a new one.